examples/retinanet/train.py

```diff
@@ -0,0 +1,529 @@
+
+  Returns:
+    A new State object with an updated `model_state` field.
+  """
+  sync_state = jax.pmap(lambda x: jax.lax.pmean(x, 'x'), 'x')
+  return meta_state.replace(model_state=sync_state(meta_state.model_state))
+
+
+def create_step_fn(lr_function):
+  """Creates a step function with a custom LR scheduler.
+
+  Args:
+    lr_function: function which takes in a single argument, the current step
+      in the training process, and yields the learning rate
+
+  Returns:
+    A function responsible with carrying out a training step. The function takes
+    in two arguments: the batch, and a `State` object, which
+    stores the current training state.
+  """
+
+  def take_step(data: Mapping[str, jnp.array],
+                meta_state: State) -> Tuple[State, Any]:
+    """Trains the model on a batch and returns the updated model.
+
+    Args:
+      data: the batch on which the pass is performed
+      meta_state: a `State` object, which holds the current model
+
+    Returns:
+      The updated model as a `State` object and the batch's loss
+    """
+
+    def _loss_fn(model: flax.nn.Model, state: flax.nn.Collection):
+      with flax.nn.stateful(state) as new_state:
+        classifications, regressions, _ = model(data['image'])
+      loss = jnp.mean(
+          retinanet_loss(classifications, regressions, data['anchor_type'],
+                         data['classification_labels'],
+                         data['regression_targets']))
+
+      return loss, new_state
+
+    # flax.struct.dataclass is immutable, so unwrap it
+    step = meta_state.step + 1
+
+    # Compute the gradients and the loss, then average them across devices
+    aux, grads = jax.value_and_grad(
+        _loss_fn, has_aux=True)(meta_state.optimizer.target,
+                                meta_state.model_state)
+    loss, new_model_state = aux
+
+    grads = jax.lax.pmean(grads, axis_name='batch')
+    loss = jax.lax.pmean(loss, axis_name='batch')
+
+    # Apply the gradients to the model
+    updated_optimizer = meta_state.optimizer.apply_gradient(
+        grads, learning_rate=lr_function(step))
+
+    # Update the meta_state
+    meta_state = meta_state.replace(
+        step=step, model_state=new_model_state, optimizer=updated_optimizer)
+
+    return meta_state, {"retinanet_loss": loss}
+
+  return take_step
+
+
+def eval_to_tensorboard(writer, evals, step, train=True, aggregate=True):
+  if aggregate:
+    evals = common_utils.get_metrics(evals)
+    summary = jax.tree_map(lambda x: x.mean(), evals)
+    logging.info("(Training Step #%d) Aggregated Metrics: %s", step, summary)
+
+  for key, vals in evals.items():
+    tag = f'{"train" if train else "eval"}_{key}'
+
+    if not isinstance(vals, (list, np.ndarray)):
+      vals = [vals]
+
+    for i, val in enumerate(vals):
+      writer.scalar(tag, val, step - len(vals) + i + 1)
+  writer.flush()
+
+
+def train_and_evaluate(config: ml_collections.ConfigDict, workdir: str) -> State:
```

Comment:
If you consider just adding a single test, then I would vote for this function (with a single train & eval step). This would make it a lot easier to catch many problems while developing. See also #176

Reply:
I haven't managed to make it happen in this PR, but will do my best to forward something in this direction in a future PR.